File: ruby/lib/SpaceStation.rb
require_relative 'SpaceStationToUI'
require_relative 'ShieldBooster'
require_relative 'Hangar'
require_relative 'CardDealer'
require_relative 'Transformation'
    
#When a attribute doesn't exists we get nil trying to get.
#On set it works as expected

module Deepspace
  class SpaceStation
    @@MAXFUEL = 100
    @@SHIELDLOSSPERUNITSHOT = 0.1
     
    attr_reader :hangar
    attr_reader :ammoPower
    attr_reader :fuelUnits
    attr_reader :name
    attr_reader :nMedals
    attr_reader :pendingDamage
    attr_reader :shieldBoosters
    attr_reader :weapons
    attr_reader :shieldPower     
    attr_reader :supplies
    
    attr_writer :pendingDamage
    attr_writer :loot
    
    def initialize(n, supplies)
      @name = n
      @supplies = supplies
      @nMedals = 0
      @weapons = Array.new
      @fuelUnits = 0
      @shieldBoosters = Array.new
      @ammoPower = supplies.ammoPower
      @shieldPower = supplies.shieldPower
      assignFuelValue(supplies.fuelUnits)
    end
    
    def self.newCopy(station)
      #Crear un supplies package
      SpaceStation.new(station.name, station.supplies)
    end
    
    def assignFuelValue(f)
      if f == nil
        @fuelUnits = 0
      else
        if f < @@MAXFUEL
          @fuelUnits = f
        else
          @fuelUnits == @@MAXFUEL
        end
      end
    end
    
    def cleanPendingDamage
      #We have to check if exits because if it's nil we are doing nil.hasNoEffect
      @pendingDamage = nil if @pendingDamage != nil && @pendingDamage.hasNoEffect
    end

    # =================
    # Discard things
    # =================
    
    def discardShieldBooster(i)
      size = @shieldBoosters.length
      
      if i >= 0 && i < size
        @shieldBoosters.delete_at(i)
        if @pendingDamage != nil
          @pendingDamage.discardShieldBooster
          cleanPendingDamage
        end
      end
    end
        
    def discardShieldBoosterInHangar(i)
      @hangar.removeShieldBooster(i) if hangar != nil
    end

    def discardWeapon(i)
      size = @weapons.length
      if i >= 0 && i < size
        w = @weapons.delete_at(i)
        if @pendingDamage != nil
          @pendingDamage.discardWeapon(w)
          cleanPendingDamage
        end
      end
    end
    
    def discardWeaponInHangar(i)
      @hangar.removeWeapon(i) if hangar != nil
    end

    # =================
    # Receive things
    # =================

    def fire
      factor = 1
      @weapons.each{|x|
        factor *= x.useIt
      }
      
      factor*@ammoPower
    end

    def mountShieldBooster(i)
      if @hangar != nil && i >= 0 && i < @hangar.shieldBoosters.length
        @shieldBoosters.push(ShieldBooster.newCopy(@hangar.removeShieldBooster(i)))  
      end
    end
    
    def mountWeapon(i)
      if @hangar != nil && i >= 0 && i < @hangar.weapons.length
        @weapons.push(Weapon.newCopy(@hangar.removeWeapon(i)))
      end
    end
    
    def receiveWeapon(w)
      if @hangar != nil
        @hangar.addWeapon(w)
      else
        false
      end
    end
    
    def receiveShieldBooster(s)
      if @hangar != nil
        @hangar.addShieldBooster(s)
      else
        false
      end
    end
    
    def receiveHangar(h)
      if @hangar == nil
        @hangar = Hangar.newCopy(h)
      end
    end
    
    def discardHangar
      @hangar = nil
    end
    
    def move
      if @fuelUnits - speed > 0
        @fuelUnits -= speed
      end
    end
    
    def speed
      @fuelUnits*1.0/@@MAXFUEL
    end
    
    def receiveSupplies(s)
      @fuelUnits += s.fuelUnits
      if @fuelUnits > @@MAXFUEL
        @fuelUnits = @@MAXFUEL
      end
      @shieldPower += s.shieldPower
      @ammoPower += s.ammoPower
    end
    
    def validState
      @pendingDamage == nil || @pendingDamage.hasNoEffect
    end
    
    def cleanUpMountedItems
      @shieldBoosters.delete_if{|x| x.uses == 0}
      @weapons.delete_if{|x| x.uses == 0}
    end
    
    def protection
      factor = 1
      @shieldBoosters.each { |x|
        factor *= x.useIt()
      }
      
      @shieldPower *factor
    end
    
    def receiveShot(shot)
      myProtection = protection
      
      if myProtection >= shot
        @shieldPower -= @@SHIELDLOSSPERUNITSHOT*shot
        @shieldPower = [0, @shieldPower].max
        return ShotResult::RESIST
      else
        shieldPower = 0
        return ShotResult::DONOTRESIST
      end 
    end
    
    def setLoot(loot)
      dealer = CardDealer.instance
      h = loot.nHangars
      
      if h > 0
        @hangar = dealer.nextHangar
        receiveHangar(dealer.nextHangar)
      end
      
      (0...loot.nSupplies).each { |s|
        receiveSupplies(dealer.nextSuppliesPackage)
      }
      
      (0...loot.nWeapons).each { |s|
        receiveWeapon(dealer.nextWeapon)
      }
      
      (0...loot.nShields).each { |s|
        receiveShieldBooster(dealer.nextShieldBooster)
      }
      
      @nMedals += loot.nMedals
      
      if loot.efficient
        Transoformation::GETEFFICIENT
      elsif loot.spaceCity  
        Transoformation::SPACECITY
      else  
        Transformation::NOTRANSFORM
      end
      
    end
         
    def setPendingDamage(d)
      @pendingDamage = d.adjust(@weapons,@shieldBoosters)
    end
    
    def getUIversion
      SpaceStationToUI.new(self)
    end
    
    def to_s
      out="Space Station:\n-Name: #{@name}"
      out+="\n-Medals: #{@nMedals}, Fuel units: #{@fuelUnits.round(2)}, Power: #{@ammoPower}, Shields: #{@shieldPower}"
      out+="\n-Weapons: [#{@weapons.join(', ')}]\n"
      out+="\n-ShieldBooster: [#{@shieldBoosters.join(', ')}]\n"
      out+="\n-Hangars: #{@hangar}\n"
      out+="\n-PendingDamage: #{@pendingDamage}\n" 
      out+="------- end of Space Station >> #{@name} << -------"
      return out
    end
  
    private :assignFuelValue, :cleanPendingDamage
  end
end

File: java/src/deepspace/SpaceStation.java
package deepspace;

import java.util.ArrayList;

class SpaceStation{

    private static final float MAXFUEL = 100;
    private static final double SHIELDLOSSPERUNITSHOT = 0.1;

    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;

    private Damage pendingDamage;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;
    private SuppliesPackage supplies;

    SpaceStation(String n, SuppliesPackage s){
	name = n;
	supplies = s;
	ammoPower = s.getAmmoPower();
	shieldPower = s.getShieldPower();
	assignFuelValue(s.getFuelUnits());

	nMedals = 0;
	hangar = null;
	weapons = new ArrayList<>();
	shieldBoosters = new ArrayList<>();
	pendingDamage = null;
    }

    private void assignFuelValue(float f){
	if (f < MAXFUEL)
	    fuelUnits = f;
	else
	    fuelUnits = MAXFUEL;
    }

    private void cleanPendingDamage(){
	if (pendingDamage != null && pendingDamage.hasNoEffect())
	    pendingDamage = null;
    }
    
    public void cleanUpMountedItems(){
        shieldBoosters.removeIf(x->x.getUses() == 0);
        shieldBoosters.removeIf(x->x.getUses() == 0);
    }

    // =================
    // Discard things
    // =================
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public int getNMedals(){
        return nMedals;
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public String getName(){
        return name;
    }
    
    public Hangar getHangar(){
        return hangar;
    }
    
    public float getFuelUnits(){
        return fuelUnits;
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public Damage getPendingDamage(){
        return pendingDamage;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public ArrayList<ShieldBooster> getShieldBooster(){
        return shieldBoosters;
    }
    
    
    public void discardHangar(){
	hangar = null;
    }

    public void discardShieldBooster(int i){
        if (i >= 0 && i < shieldBoosters.size()){
	    shieldBoosters.remove(i);
	    if (pendingDamage != null){
		pendingDamage.discardShieldBooster();
		cleanPendingDamage();
	    }
	    
	}
    }

    public void discardShieldBoosterInHangar(int i){
	if (hangar != null)
	    hangar.removeShieldBooster(i);
    }

    public void discardWeapon(int i){
	if (i >= 0 && i < weapons.size()){
	    Weapon w = weapons.remove(i);
	    if (pendingDamage != null){
		pendingDamage.discardWeapon(w);
		cleanPendingDamage();
	    }
	    
	}
    }

    public void discardWeaponInHangar(int i){
	if (hangar != null)
	    hangar.removeWeapon(i);
    }

    public float fire(){
	float factor = 1;
	for(Weapon w: weapons){
	    factor *= w.useIt(); 
	}

	return factor*ammoPower;
	
    }

    // =================
    // getters
    // =================
    
    public int getNMedals(){
        return nMedals;
    }
    

    
    public float getSpeed(){
        return fuelUnits/MAXFUEL;
    }

    public void mountShieldBooster(int i){
	if (hangar != null && i >= 0 && i < hangar.getShieldBoosters().size()){
	    shieldBoosters.add(new ShieldBooster(hangar.removeShieldBooster(i)));
	}
    }

    public void mountWeapon(int i){
	if (hangar != null && i >= 0 && i < hangar.getWeapons().size()){
	    weapons.add(new Weapon(hangar.removeWeapon(i)));
	}
    }

    public void move(){
        if (fuelUnits - getSpeed() > 0)
            fuelUnits -= getSpeed();
    }
    
    public float protection(){
        float factor = 1;
        
        for(ShieldBooster s: shieldBoosters){
	    factor *= s.useIt(); 
	}

	return factor*shieldPower;
    }
    
    
    // =================
    // Receive things
    // =================

    public void receiveHangar(Hangar h){
        if (hangar == null)
            hangar = new Hangar(h);
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        if (hangar != null)
            return hangar.addShieldBooster(s);
        else
            return false;
    }
    
    public ShotResult receiveShot(float shot){
        float myProtection = protection();
        
        if (myProtection >= shot){
            shieldPower -= SHIELDLOSSPERUNITSHOT*shot;
            shieldPower = Math.max(0, shieldPower);
            return ShotResult.RESIST;
        }
        else{
            shieldPower = 0;
            return ShotResult.DONOTRESIST;
        }        
    }
    
    public void receiveSupplies(SuppliesPackage s){
        fuelUnits += s.getFuelUnits();
        if (fuelUnits > MAXFUEL)
            fuelUnits = MAXFUEL;
        
        shieldPower += s.getShieldPower();
        ammoPower += s.getAmmoPower();
    }
    
    public boolean receiveWeapon(Weapon w){
        if (hangar != null)
            return hangar.addWeapon(w);
        else
            return false;
    }
    
    public void setLoot(Loot loot){
        CardDealer dealer = CardDealer.getInstance();
        
        int h = loot.getNHangars();
            
        if (h > 0){
            hangar = dealer.nextHangar();
            receiveHangar(dealer.nextHangar());
        }
        
        for(int i = 0; i < loot.getNSupplies(); i++){
            receiveSupplies(dealer.nextSuppliesPackage());
        }
        
        for(int i = 0; i < loot.getNShields(); i++){
            receiveShieldBooster(dealer.nextShieldBooster());
        }
        
        for(int i = 0; i < loot.getNWeapons(); i++){
            receiveWeapon(dealer.nextWeapon());
        }
        
        nMedals += loot.getNMedals();   
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public boolean validState(){
        return pendingDamage == null || pendingDamage.hasNoEffect();
    }
    
}
